relayer/relays/parachain/catchup.go

```diff
@@ -192,8 +190,10 @@ func (li *BeefyListener) parablocksWithProofs(blocks []ParaBlockWithDigest,
 		}
 
 		// Note - relayChainBlockNumber will be one less than the actual block number we want
-		// due to the decrement at the end of the loop, so we add back 1
-		mmrProof, err := li.relaychainConn.GetMMRLeafForBlock(uint64(relayChainBlockNumber+1), latestRelaychainBlockHash, li.config.Polkadot.BeefyStartingBlock)
+		// due to the decrement at the end of the loop, so we increment by 1. Additionally,
+		// parachain merkle roots are created 1 block later than the actual parachain headers,
+		// so we increment twice.
```

Comment:
Are you sure this is stable? Beefy runs async between validators, and IIUC it could lag more than 1 block behind parachain heads being committed.

Reply:
@philipstanislaus Beefy runs async, so could lag behind, but once it gets done it should still produce one leaf for every block and so the function that maps from mmrLeafIndex -> mmrLeaf -> paraHeadsRoot -> Relaychain block number which commits to those Paraheads should still be constant. If not, then it means the MMR could be missing some para heads?

Also, if beefy is lagging far behind, the call to GenerateProof will fail and the relayer will just continue and try again later just fine :).

Having said that, I'm not sure of the above - it includes a bunch of assumptions that we need to dig in to verify, including:
 - The polkadot relay chain does finalize every single parachain head, and include every head of every parachain in Paras.Head over time. It never skips a parachain head or finalizes multiple heads from a single parachain in a row in a single relay chain block
 - The same for beefy - Beefy MMR leaves are 1:1 with relaychain blocks (besides for blocks created before the existence of beefy) and there is one beefy leaf for every relaychain block and therefore one beefy leaf for the Paras.Heads value at every relaychain block. There is thus effectively an isomorphism from beefy leaf -> Paras.Heads
 - This isomorphism is a straightforward linear function, specifically:
Paras.Heads(relayChainBlockNumber) = Paras.Heads(mmrLeafIndex)

Reply:
Did just dive into the parachain consensus code (https://github.com/paritytech/cumulus/blob/master/client/consensus/common/src/parachain_consensus.rs) a bit. 

Collators do not create blocks, they create block candidates. Actual block production and finalization is done by the assigned validators of the relay chain. I'm still not sure whether there is any guarantee that every single finalized parachain block is included in a MMR leaf's parachain root. It seems possible that parachain blocks are skipped and never included in that root. 

I'll try to find out whether the hypothesis that parachain blocks can be skipped holds or not.

I can confirm though that there is a 1:1 relationship between beefy leaves and relaychain blocks.

Reply:
@philipstanislaus cc @vgeddes hmm ok. if it is possible that parachain blocks are skipped and never included in that root, then we have a bigger problem as commitments will get lost. We'd need to move to a commitment mechanism which keeps a buffer of old commitments and/or messages over a few blocks to ensure there are no gaps. Let us know when you get deeper.

Reply:
I think it's important to differentiate between the parachain consensus provided by the relay chain and the way that framework is used in Cumulus.

What the relay chain consensus provides is that maximum one state transition is processed per relay chain block. Therefore, the head-data output by each the state transition is present in the post-state of at least one relay chain block.

The implementation of Cumulus chains means one block is one state transition. So nothing is skipped, as you would like to have as a property. But this is a Cumulus implementation detail which could change in the future by combining multiple blocks into one state transition

Reply:
@rphmeier thanks for the clarification.

What does a para head in polkadot’s the paras pallet (https://github.com/paritytech/polkadot/blob/4165c7f83a85e12c7ec56366a9559bf0eba25f8d/runtime/parachains/src/paras.rs#L328) refer to? One state transition?

Also if cumulus changes in the future, we could stay on an implementation that associates one block with one state transition, right?